core/src/Kokkos_MathematicalSpecialFunctions.hpp

```diff
@@ -1268,6 +1268,222 @@ KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_h21(const CmplxType& z) {
   return ch21;
 }
 
+template <auto BesselZero, auto BesselOne, class CmplxType, class IntType,
+          class... Args>
+KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel(const IntType& a,
+                                            const CmplxType& z, Args... args) {
+  if (a == 0) return BesselZero(z, args...);
+  if (a == 1) return BesselOne(z, args...);
+  CmplxType bessel_im2 = BesselZero(z, args...);
+  CmplxType bessel_im1 = BesselOne(z, args...);
+
+  for (IntType cur_a = 2; cur_a <= a; ++cur_a) {
+    CmplxType tmp = bessel_im1;
+    bessel_im1    = CmplxType(2 * (cur_a - 1)) * bessel_im1 / z - bessel_im2;
+    bessel_im2    = tmp;
+  }
+
+  return bessel_im1;
+}
+
+template <class CmplxType, class RealType, class IntType>
+KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_j(const IntType& a,
+                                              const CmplxType& z,
+                                              const RealType& joint_val = 25.0,
+                                              const IntType& bw_start   = 70) {
+  if ((z.real() == 0.0) && (z.imag() == 0.0)) {
+    return CmplxType(0.0, 0.0);
+  }
+  auto pos_a  = Kokkos::abs(a);
+  auto factor = a < 0 ? Kokkos::pow(-1.0, pos_a) : 1.0;
+  return factor *
+         cyl_bessel<cyl_bessel_j0<CmplxType, RealType, IntType>,
+                    cyl_bessel_j1<CmplxType, RealType, IntType>, CmplxType,
+                    IntType, RealType>(pos_a, z, joint_val, bw_start);
```

Comment:
Can't we just pass the values for `cyl_bessel_j0` and `cyl_bessel_j1` on? (Not necessarily a problem).

Reply:
previous implementation was using recursion so this functions where required
Not sure that we can change it now since it will require constexpr evaluation of `cyl_ besse{}{} ` and if we pass them by value then for zero and first order instead of one evaluation we end up with two

Reply:
I think it's fine.